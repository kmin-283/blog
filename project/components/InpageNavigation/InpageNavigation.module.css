.wrapper {
    width: 20%;
    font-size: 0.875rem;
}

.container {
    margin: 0.5rem 0;
    padding: 0.5rem 0;
    border-right: 2px solid var(--main-theme-color);
    border-left: 2px solid var(--main-theme-color);
    background-color: var(--admin-background-color);
    animation: popUp 0.3s;
}

.fixed {
    position: fixed;
    top: 0;
    right: 0;
    width: 20%;
    transition-duration: 0.25s;
    transform: translateY(50px);
}