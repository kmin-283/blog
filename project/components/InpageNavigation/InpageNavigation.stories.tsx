import React from 'react';
import { Story, Meta } from '@storybook/react';
import InpageNavigation, {InpageNavigationProps} from "@/components/InpageNavigation/InpageNavigation";

export default {
  title: 'Components/InpageNavigation',
  component: InpageNavigation,
} as Meta;

const Template: Story<InpageNavigationProps> = ({internalLinks}) => {
  return (
    <div style={{width: 800}}>
      <InpageNavigation internalLinks={internalLinks} />
    </div>);
};

export const BasicInpageNavigation = Template.bind({});
BasicInpageNavigation.args = {
  internalLinks: "[" +
    "{\"headingLevel\":2,\"value\":\"1 헤딩\",\"child\":[]}," +
    "{\"headingLevel\":2,\"value\":\"2 헤딩\",\"child\":[" +
    "{\"headingLevel\":3,\"value\":\"2-1 헤딩\",\"child\":[" +
    "{\"headingLevel\":4,\"value\":\"2-1-1 헤딩\",\"child\":[]}," +
    "{\"headingLevel\":4,\"value\":\"2-1-2 헤딩\",\"child\":[" +
    "{\"headingLevel\":5,\"value\":\"2-1-2-1 헤딩\",\"child\":[]}]}," +
    "{\"headingLevel\":4,\"value\":\"2-1-3 헤딩\",\"child\":[]}]}," +
    "{\"headingLevel\":3,\"value\":\"2-2 헤딩\",\"child\":[]}]}]\n"
};