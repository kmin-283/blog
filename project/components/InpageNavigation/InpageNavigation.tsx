import React, {FC} from 'react';
import {markedString} from "@/utils/markdown";
import styles from './InpageNavigation.module.css';

interface ContentHeaderProps {
  internalLinks: string[];
}

const InpageNavigation: FC<ContentHeaderProps> = ({internalLinks}) => {
  return (
    <nav className={styles.container}>
      <ol className={styles.links}>
        {internalLinks.map((internalLink, index) =>
          (internalLink &&
            <li
              className={styles.link}
              key={internalLink}
              dangerouslySetInnerHTML={{
                __html: markedString(`[${index + 1}. ${internalLink
                  .replace(/-/g, ' ')}](#${internalLink})`)
              }}>
            </li>))}
      </ol>
    </nav>
  );
};

export default InpageNavigation;