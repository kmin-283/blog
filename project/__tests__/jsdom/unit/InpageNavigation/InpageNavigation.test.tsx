import React from "react";
import renderer from "react-test-renderer";
import InpageNavigation from "@/components/InpageNavigation/InpageNavigation";
import { render, cleanup, RenderResult } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

describe("InpageNavigation 컴포넌트", () => {
  const internalLinks =
    "[" +
    '{"headingLevel":2,"value":"1 헤딩","child":[]},' +
    '{"headingLevel":2,"value":"2 헤딩","child":[' +
    '{"headingLevel":3,"value":"2-1 헤딩","child":[' +
    '{"headingLevel":4,"value":"2-1-1 헤딩","child":[]},' +
    '{"headingLevel":4,"value":"2-1-2 헤딩","child":[' +
    '{"headingLevel":5,"value":"2-1-2-1 헤딩","child":[]}]},' +
    '{"headingLevel":4,"value":"2-1-3 헤딩","child":[]}]},' +
    '{"headingLevel":3,"value":"2-2 헤딩","child":[]}]}]\n';
  let component: RenderResult;
  afterAll(cleanup);
  beforeEach(() => {
    component = render(<InpageNavigation internalLinks={internalLinks} />);
  });

  test("렌더링", () => {
    const inpageNavigation = renderer
      .create(<InpageNavigation internalLinks={internalLinks} />)
      .toJSON();
    expect(inpageNavigation).toMatchSnapshot();
  });

  test("내부 링크 테스트", () => {
    const { getAllByRole } = component;
    const listItems = getAllByRole("listitem");
    const orderedLists = getAllByRole("list");

    expect(orderedLists).toHaveLength(4);
    expect(listItems).toHaveLength(8);
  });

  test("내부 링크는 계층 구조를 갖는다", () => {
    const { getByText } = component;
    const h3 = getByText("2-1 헤딩");
    const h3Children = h3.nextSibling! as HTMLOListElement;

    if (h3Children) {
      // [
      // <li><a href='/#2-1-1-%ED%97%A4%EB%94%A9'>2-1-1 헤딩</a></li>,
      // <li>
      //  <a href='/#2-1-2-%ED%97%A4%EB%94%A9'>2-1-2 헤딩</a>
      //  <ol class='links'>
      //    <li><a href='/#2-1-2-1-%ED%97%A4%EB%94%A9'>2-1-2-1 헤딩</a></li>
      //   </ol>
      //  </li>,
      //  <li><a href='/#2-1-3-%ED%97%A4%EB%94%A9'>2-1-3 헤딩</a></li>
      //  ]
      expect(h3Children.children).toHaveLength(3);

      if (h3Children.children.length > 0) {
        const grandChildren = h3Children.children[1].querySelector("ol");
        expect(grandChildren?.children).toHaveLength(1);
      }
    }
  });
});
