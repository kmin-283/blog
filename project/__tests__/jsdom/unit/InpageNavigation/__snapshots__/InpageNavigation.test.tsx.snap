// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InpageNavigation 컴포넌트 렌더링 1`] = `
<section
  className="wrapper"
>
  <nav
    className="container "
  >
    <ol
      className="links"
    >
      <li>
        <a
          href="/#1-%ED%97%A4%EB%94%A9"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          1 헤딩
        </a>
      </li>
      <li>
        <a
          href="/#2-%ED%97%A4%EB%94%A9"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          2 헤딩
        </a>
        <ol
          className="links"
        >
          <li>
            <a
              href="/#2-1-%ED%97%A4%EB%94%A9"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              2-1 헤딩
            </a>
            <ol
              className="links"
            >
              <li>
                <a
                  href="/#2-1-1-%ED%97%A4%EB%94%A9"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                >
                  2-1-1 헤딩
                </a>
              </li>
              <li>
                <a
                  href="/#2-1-2-%ED%97%A4%EB%94%A9"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                >
                  2-1-2 헤딩
                </a>
                <ol
                  className="links"
                >
                  <li>
                    <a
                      href="/#2-1-2-1-%ED%97%A4%EB%94%A9"
                      onClick={[Function]}
                      onMouseEnter={[Function]}
                    >
                      2-1-2-1 헤딩
                    </a>
                  </li>
                </ol>
              </li>
              <li>
                <a
                  href="/#2-1-3-%ED%97%A4%EB%94%A9"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                >
                  2-1-3 헤딩
                </a>
              </li>
            </ol>
          </li>
          <li>
            <a
              href="/#2-2-%ED%97%A4%EB%94%A9"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              2-2 헤딩
            </a>
          </li>
        </ol>
      </li>
    </ol>
  </nav>
</section>
`;
